File: solutions/18/main3.cpp
#include <iostream>
#include <unordered_map> 
#include <algorithm>
#include <numeric> 
#include <gtest/gtest.h>

using namespace std; 

// use multimap
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        
        vector<vector<int>> result; 
        
        // check input 
        if (nums.size() < 4)
            return result; 
        
        sort(nums.begin(), nums.end());
        
        unordered_multimap<int, pair<int, int>> cache; 
        
        for (size_t i = 0; i < nums.size() - 1; ++i)
        {
            for (size_t j = i + 1; j < nums.size(); ++j)
            {
                cache.insert(make_pair(nums[i] + nums[j], make_pair(i, j)));
            }
        }
                             
        for (auto i = cache.begin(); i != cache.end(); ++i)
        {
            int diff = target - i->first; 
            auto range = cache.equal_range(diff);
            for (auto j = range.first; j != range.second; ++j)
            {
                auto a = i->second.first; 
                auto b = i->second.second; 
                auto c = j->second.first; 
                auto d = j->second.second; 
                if (a != c && a != d && b != c && b != d) 
                {
                    vector<int> vec = {nums[a], nums[b], nums[c], nums[d]};
                    sort(vec.begin(), vec.end());
                    result.push_back(vec);
                }
            }
        }
                             
        sort(result.begin(), result.end());
        result.erase(unique(result.begin(), result.end()), result.end());
        return result; 
    }
};

TEST(Test18, SimpleTest)
{
    vector<int> nums = {1,0,-1,0,-2,2};
    const int target = 0; 
    Solution s; 
    vector<vector<int>> result = s.fourSum(nums, target); 
    for (const auto r : result)
    {
	EXPECT_EQ(accumulate(r.begin(), r.end(), 0), target) << "Wrong answer"; 
    }
}

int main() 
{
    testing::InitGoogleTest(); 
    return RUN_ALL_TESTS(); 
}

File: solutions/719/main3.py

# wrong 
import heapq 
class Solution(object):
    def smallestDistancePair(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        pq = []
        nums.sort()
        n = len(nums)
        for i in range(1, n):
            heapq.heappush(pq, (nums[i]-nums[i-1], i-1, i))
        while len(pq) > 0: 
            diff, i, j = heapq.heappop(pq) 
            k -= 1 
            if k == 0: 
                return diff
            if j < n-1:  
                heapq.heappush(pq, (nums[j+1]-nums[i], i, j+1))
            if i > 0:
                heapq.heappush(pq, (nums[j]-nums[i-1], i-1, j))
        
if __name__ == "__main__": 
    s = Solution()

    assert s.smallestDistancePair([1,3,1], 1) == 0

    assert s.smallestDistancePair([1,6,1], 3) == 5

    assert s.smallestDistancePair([62,100,4], 2) == 58